### PR TITLE
Properly delete factorization of SuperLURowLocMatrix

### LINKED ISSUE #2368: Memory leak in example ex1p with superLU
Dear MFEM community, 

I found memory leak in my MFEM-based solver and reproduced it in **examples/superlu/ex1p.cpp**. I slightly modified it, add an outer loop in order to call SuperLU solver several times:
```
 // 13. Solve the linear system A X = B utilizing SuperLU.
   // Add this loop to check for memory leakage. If there is a leak wi will see it 
   // in "top" command output
   for (int i = 0; i < 5; i++){
   // 13. Solve the linear system A X = B utilizing SuperLU.
   // Add this loop to check for memory leakage. If there is a leak wi will see it 
   // in "top" command output
   for (int i = 0; i < 5; i++){
      SuperLUSolver *superlu = new SuperLUSolver(MPI_COMM_WORLD);
      Operator *SLU_A = new SuperLURowLocMatrix(*A.As<HypreParMatrix>());
      superlu->SetPrintStatistics(true);
      superlu->SetSymmetricPattern(false);
...
```
I've also added ```delete``` for all pointers inside the loop.
The remaining code is untouched. When I run the program I see in the output of **top** command that used memory is increasing with each loop iteration.
I have also checked the program with **valgrind**, it detects leak and points to ```superlu->Mult(B, X);```, but I am not accustomed to valgrind language and can't interpret it properly)
However, from the size of the memory which is added at each iteration it is clear that the memory for LU factors are not freed. 

Could you please check it? May be we can check somehow whether the problem is on the MFEM side or inside the SuperLU?

Thanks in advance,
Aleksandr Chikitkin

### PR DESCRIPTION
**Pull request by jandrej**
Resolves #2368

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2420](https://github.com/mfem/mfem/pull/2420) | @jandrej | @tzanio | @psocratis + @tzanio | 07/21/21 | 07/21/21 | 07/22/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
How does this relate to #2419 (and #2413)?

File: linalg/superlu.cpp [hunk 1]
```diff
@@ -189,6 +189,9 @@ SuperLURowLocMatrix::SuperLURowLocMatrix( const HypreParMatrix & hypParMat )
    dCreate_CompRowLoc_Matrix_dist(A, m, n, nnz_loc, m_loc, fst_row,
                                   nzval, colind, rowptr,
                                   SLU_NR_loc, SLU_D, SLU_GE);
+
+   // Save global number of columns (width) of the matrix
+   num_global_cols = m;
```
Comment (psocratis) [inline linalg/superlu.cpp:194]: this should be `n`, shouldn't be?

**Comment by psocratis** [comment]
> How does this relate to #2419 (and #2413)?

They should be merged, i think

File: linalg/superlu.cpp (hunk 1 above)
Comment (jandrej) [reply linalg/superlu.cpp:194]: Argh, yes.

**Comment by tzanio** [comment]
> They should be merged, i think

@psocratis, can you do that? It will be better to review a single PR

**Comment by psocratis** [comment]
> @psocratis, can you do that? It will be better to review a single PR

I will. I am just doing some testing right now

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
LGTM

**Review by psocratis — APPROVED** [review]

**Comment by psocratis** [comment]
@tzanio, thanks!

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by pghysels — APPROVED** [review]

**Comment by cdelv** [comment]
@psocratis @jandrej 
Apparently is fixed. Example 1 with 100 loops
```
	Command being timed: "mpirun -np 4 ./a.out"
	User time (seconds): 20.55
	System time (seconds): 0.30
	Percent of CPU this job got: 376%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:05.53
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 38508
```
with 1000 loops
```
	Command being timed: "mpirun -np 4 ./a.out"
	User time (seconds): 257.30
	System time (seconds): 2.11
	Percent of CPU this job got: 398%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 1:05.17
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 38956
```
How do I approve?

**Comment by psocratis** [comment]
@cdelv, you can press the tab `Files Changed` then on the right you will see `Review Changes` and from there you can approve. 

Thanks,

Socratis

**Review by cdelv — APPROVED** [review]

**Pull request merged**